### PR TITLE
Adding time series anomaly detection code example.

### PR DESCRIPTION
**Pull request by pavithrasv**
Adding time series anomaly detection code example - using convolutional autoencoder.

### CODE REVIEW

File: examples/structured_data/time_series_anomaly_using_autoencoder.py [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+Date created: 2020/05/31
+Last modified: 2020/05/31
+Description: Detect anomalies in a timeseries using an Autoencoder.
+"""
+
+"""
+## Introduction
+
+This script demonstrates how you can use a reconstruction convolutional autoencoder model
+to detect anomolies in timeseries data.
```
Comment (fchollet) [inline examples/structured_data/time_series_anomaly_using_autoencoder.py:13]: anomalies

File: examples/structured_data/time_series_anomaly_using_autoencoder.py [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+
+We train the model on timeseries data which does not have any anomalies. We will use the
+[Numenta Anomaly Benchmark](https://www.kaggle.com/boltzmannbrain/nab?) dataset to
```
Comment (fchollet) [inline examples/structured_data/time_series_anomaly_using_autoencoder.py:16]: No `?` in url; please describe the dataset used to provide context: what does the data represent? why is this a good fit for demonstrating anomaly detection?

File: examples/structured_data/time_series_anomaly_using_autoencoder.py [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+Author: [pavithrasv](https://github.com/pavithrasv)
+Date created: 2020/05/31
+Last modified: 2020/05/31
+Description: Detect anomalies in a timeseries using an Autoencoder.
+"""
+
+"""
+## Introduction
+
+This script demonstrates how you can use a reconstruction convolutional autoencoder model
```
Comment (fchollet) [inline examples/structured_data/time_series_anomaly_using_autoencoder.py:12]: Is there a reference you can provide for this model?

File: examples/structured_data/time_series_anomaly_using_autoencoder.py [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+seed(RANDOM_SEED)
+tf.random.set_seed(RANDOM_SEED)
+
+"""
+## Load the data
+
+We will use the  dataset [Numenta Anomaly Benchmark
+(NAB)](https://www.kaggle.com/boltzmannbrain/nab?) dataset. We will use the
+`art_daily_small_noise.csv` file for training and the `art_daily_jumpsup.csv` file for
+testing anomaly detection.
```
Comment (fchollet) [inline examples/structured_data/time_series_anomaly_using_autoencoder.py:52]: Likewise, we should provide more context. Read your example from the perspective of a first-time reader.

File: examples/structured_data/time_series_anomaly_using_autoencoder.py [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+
+from datetime import datetime
+from matplotlib import pyplot as plt
+from matplotlib import dates as md
+from numpy.random import seed
+from tensorflow.keras import layers
+from tensorflow.keras import Sequential
+
+"""
+## Set random seed
```
Comment (fchollet) [inline examples/structured_data/time_series_anomaly_using_autoencoder.py:39]: You don't need this in a code example

File: examples/structured_data/time_series_anomaly_using_autoencoder.py [hunk 6]
```diff
@@ -0,0 +1,344 @@
...
+    output = []
+    for i in range(len(values) - time_steps):
+        output.append(values[i : (i + time_steps)])
+    # Convert 2D sequences into 3D as we will be feeding this into
+    # a convolutional layer.
+    return np.expand_dims(output, axis=2)
+
+
+x_train = create_sequences(training_value)
+print(x_train.shape)
```
Comment (fchollet) [inline examples/structured_data/time_series_anomaly_using_autoencoder.py:155]: Please add a complete log message here

File: examples/structured_data/time_series_anomaly_using_autoencoder.py [hunk 7]
```diff
@@ -0,0 +1,344 @@
...
+shape. In this case, `sequence_length` is 288 and `num_features` is 1.
+"""
+
+model = Sequential(
+    [
+        layers.Conv1D(
+            filters=32,
+            kernel_size=7,
+            padding="same",
+            input_shape=(x_train.shape[1], x_train.shape[2]),
```
Comment (fchollet) [inline examples/structured_data/time_series_anomaly_using_autoencoder.py:171]: Rather than using `input_shape` prefer starting the model with an `Input()`

File: examples/structured_data/time_series_anomaly_using_autoencoder.py [hunk 8]
```diff
@@ -0,0 +1,344 @@
...
+        ),
+        layers.MaxPool1D(padding="same"),
+        layers.Dropout(rate=0.2),
+        layers.Conv1D(filters=16, kernel_size=7, padding="same"),
+        layers.MaxPool1D(padding="same"),
+        layers.Conv1D(filters=16, kernel_size=7, padding="same"),
+        layers.UpSampling1D(),
```
Comment (fchollet) [inline examples/structured_data/time_series_anomaly_using_autoencoder.py:178]: Since the model is meant to be symmetrical I recommend using `Conv1D` for the first half and `Conv1DTranspose` for the second half.

Rather than `MaxPool1D` and `UpSampling1D`, consider using strides in the convolutions. Maxpooling a an operations that is hard to reverse; it works great in classification settings but should be avoided in an autoencoder.

File: examples/structured_data/time_series_anomaly_using_autoencoder.py [hunk 9]
```diff
@@ -0,0 +1,344 @@
...
+        layers.Dropout(rate=0.2),
+        layers.Conv1D(filters=32, kernel_size=7, padding="same"),
+        layers.UpSampling1D(),
+        layers.Conv1D(filters=1, kernel_size=7, padding="same"),
+    ]
+)
+model.compile(optimizer=tf.keras.optimizers.Adam(learning_rate=0.001), loss="mae")
```
Comment (fchollet) [inline examples/structured_data/time_series_anomaly_using_autoencoder.py:185]: MAE is generally not a good loss because it lacks smoothness. MSE is almost universally better for this use case. You can track MAE as a metric.

File: examples/structured_data/time_series_anomaly_using_autoencoder.py [hunk 10]
```diff
@@ -0,0 +1,344 @@
...
+
+Please note that we are using `x_train` as both the input and the target since this is a
+reconstruction model.
+"""
+
+history = model.fit(
+    x_train,
+    x_train,
+    epochs=50,
+    batch_size=100,
```
Comment (fchollet) [inline examples/structured_data/time_series_anomaly_using_autoencoder.py:199]: Generally, use powers of 2 for the batch size, to make GPU memory allocation easier. Here, e.g. 128

File: examples/structured_data/time_series_anomaly_using_autoencoder.py (hunk 1 above)
Comment (pavithrasv) [reply examples/structured_data/time_series_anomaly_using_autoencoder.py:13]: Done.

File: examples/structured_data/time_series_anomaly_using_autoencoder.py (hunk 5 above)
Comment (pavithrasv) [reply examples/structured_data/time_series_anomaly_using_autoencoder.py:39]: Done.

File: examples/structured_data/time_series_anomaly_using_autoencoder.py (hunk 6 above)
Comment (pavithrasv) [reply examples/structured_data/time_series_anomaly_using_autoencoder.py:155]: Done.

File: examples/structured_data/time_series_anomaly_using_autoencoder.py (hunk 7 above)
Comment (pavithrasv) [reply examples/structured_data/time_series_anomaly_using_autoencoder.py:171]: Done.

File: examples/structured_data/time_series_anomaly_using_autoencoder.py (hunk 9 above)
Comment (pavithrasv) [reply examples/structured_data/time_series_anomaly_using_autoencoder.py:185]: Done.

File: examples/structured_data/time_series_anomaly_using_autoencoder.py (hunk 10 above)
Comment (pavithrasv) [reply examples/structured_data/time_series_anomaly_using_autoencoder.py:199]: Done.

File: examples/structured_data/time_series_anomaly_using_autoencoder.py (hunk 2 above)
Comment (pavithrasv) [reply examples/structured_data/time_series_anomaly_using_autoencoder.py:16]: Removed this text form here and added more explanation in the load data section below.

File: examples/structured_data/time_series_anomaly_using_autoencoder.py (hunk 4 above)
Comment (pavithrasv) [reply examples/structured_data/time_series_anomaly_using_autoencoder.py:52]: Done.

File: examples/structured_data/time_series_anomaly_using_autoencoder.py (hunk 3 above)
Comment (pavithrasv) [reply examples/structured_data/time_series_anomaly_using_autoencoder.py:12]: I wanted to tryout timeseries anomaly detection, when reading about it i saw that autoencoders were used in the supervised solutions for this problem. I then played with LSTM/Conv1D architecture and arrived at this.

File: examples/structured_data/time_series_anomaly_using_autoencoder.py (hunk 8 above)
Comment (pavithrasv) [reply examples/structured_data/time_series_anomaly_using_autoencoder.py:178]: Used `Conv1DTranspose` and replaced `MaxPool1D` with strides in `Conv1D` layers. 

Not sure how to replace `UpSampling1D`.

**Comment by pavithrasv** [comment]
Thank you for the review. Addressed comments, PTAL.

File: examples/structured_data/time_series_anomaly_using_autoencoder.py (hunk 8 above)
Comment (fchollet) [reply examples/structured_data/time_series_anomaly_using_autoencoder.py:159]: You should be able to use `strides` in the `Conv1DTranspose` layers to achieve this.

Comment (pavithrasv) [reply examples/structured_data/time_series_anomaly_using_autoencoder.py]: I tried it but thought that model did not detect anomalies given the same parameters.. must have been some glitch. I tried it again now and it worked.

File: examples/structured_data/time_series_anomaly_using_autoencoder.py [hunk 11]
```diff
@@ -0,0 +1,323 @@
...
+
+"""
+## Build a model
+
+We will build a convolutional reconstruction autoencoder model. The model will take input
+of shape `(batch_size, sequence_length, num_features)` and return output of the same
+shape. In this case, `sequence_length` is 288 and `num_features` is 1.
+"""
+
+model = Sequential(
```
Comment (fchollet) [inline examples/structured_data/time_series_anomaly_using_autoencoder.py:152]: I would also recommend adding `relu` activations in the intermediate layers (not the last layer). The lack of non-linearity here is surprising

File: examples/structured_data/time_series_anomaly_using_autoencoder.py [hunk 12]
```diff
@@ -0,0 +1,323 @@
...
+from matplotlib import pyplot as plt
+from matplotlib import dates as md
+from numpy.random import seed
+from tensorflow.keras import layers
+from tensorflow.keras import Sequential
+
+"""
+## Load the data
+
+We will use the [Numenta Anomaly Benchmark(NAB)](https://www.kaggle.com/boltzmannbrain/nab) dataset. It provides artifical timeseries data containing labeled anomalous periods of behavior. Data are ordered, timestamped, single-valued metrics.
```
Comment (fchollet) [inline examples/structured_data/time_series_anomaly_using_autoencoder.py:37]: Please shorten text lines to ~80 chars (for version control)

Comment (pavithrasv) [reply examples/structured_data/time_series_anomaly_using_autoencoder.py:37]: Done.

File: examples/structured_data/time_series_anomaly_using_autoencoder.py (hunk 11 above)
Comment (pavithrasv) [reply examples/structured_data/time_series_anomaly_using_autoencoder.py:160]: Yes, it was surprising to me as well. Anything implicit Keras is doing that is non-linear? I added relu activations now.

Comment (fchollet) [reply examples/structured_data/time_series_anomaly_using_autoencoder.py:160]: Yes, the strides and the mixing across convolution windows as you stack multiple layers make the transformations non-linear. However relu should definitely help with learning.
